lib/src/macro.dart

```diff
@@ -4,44 +4,43 @@ import 'types.dart';
 
 abstract class Macro {}
 
-abstract class TypeMacro implements Macro {}
-
-abstract class DeclarationMacro implements Macro {}
-
-abstract class DefinitionMacro implements Macro {}
-
 abstract class ClassTypeMacro implements TypeMacro {
-  void type(TargetClassType type);
+  void type(ClassType type, TypeBuilder builder);
 }
 
 abstract class ClassDeclarationMacro implements DeclarationMacro {
-  void declare(TargetClassDeclaration declaration);
+  void declare(ClassDeclaration declaration, ClassDeclarationBuilder builder);
 }
 
-abstract class ClassDefinitionMacro implements DefinitionMacro {
-  void define(TargetClassDefinition definition);
-}
+// This isn't needed. Instead, a class-level macro that adds members
+// must implement FieldDefinitionMacro and/or MethodDefinitionMacro and then
+// has those invoked directly for the members it has declared.
+//
+// A macro that declares a member is committing itself to defining it later.
+// abstract class ClassDefinitionMacro implements DefinitionMacro {
+//   void define(TargetClassDefinition definition);
+// }
```

Comment:
I think this is probably OK, but there might be some use cases that it doesn't necessarily support well.

Specifically a class level macro wouldn't be able to visit all the _existing_ declarations in that class and do operations on them. For instance imagine a class level macro that adds some wrapper code to every single method in the class, how would it do that?

Reply:
Another idea here is if we separated the thing that is annotated from the things that are visited a bit more. So for instance you could have a macro that only implements `FieldDefinitionMacro`, but you are allowed to annotate the class or individual fields. If you annotate the class then it calls `visitFieldDefinition` for all fields in the class, automatically.